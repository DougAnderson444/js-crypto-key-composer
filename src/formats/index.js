import * as pkcs1Der from './pkcs1/pkcs1-der';
import * as pkcs1Pem from './pkcs1/pkcs1-pem';
import * as pkcs8Der from './pkcs8/pkcs8-der';
import * as pkcs8Pem from './pkcs8/pkcs8-pem';
import * as spkiDer from './spki/spki-der';

export { pkcs1Der, pkcs1Pem, pkcs8Der, pkcs8Pem, spkiDer };
