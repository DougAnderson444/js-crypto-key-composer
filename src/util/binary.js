import { Buffer } from 'buffer';

export const binaryStringToArrayBuffer = (str) => {
    const len = str.length;
    const uint8Array = new Uint8Array(len);

    for (let i = 0; i < len; i += 1) {
        uint8Array[i] = str.charCodeAt(i);
    }

    return uint8Array.buffer;
};

export const arrayBufferToBinaryString = (arrayBuffer) =>
    String.fromCharCode.apply(null, new Uint8Array(arrayBuffer));

export const hexStringToArrayBuffer = (str) =>
    new Uint8Array(str.match(/.{1,2}/g).map((byte) => parseInt(byte, 16))).buffer;

export const arrayBufferToHexString = (arrayBuffer) =>
    Array.prototype.map.call(new Uint8Array(arrayBuffer), (x) => (`00${x.toString(16)}`).slice(-2)).join('');

export const nodeBufferToArrayBuffer = (buffer) =>
    buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);

export const arrayBufferToNodeBuffer = (arrayBuffer) => Buffer.from(arrayBuffer);

export const bnToArrayBuffer = (bn) => {
    const numArray = bn.toArray();

    /* eslint-disable no-bitwise */
    // Remove useless sign
    if (!bn.negative && numArray[0] & 0x80) {
        numArray.unshift(0);
    }
    /* eslint-enable no-bitwise */

    return Uint8Array.from(numArray).buffer;
};

export const integerFromArrayBuffer = (arrayBuffer) => {
    if (arrayBuffer.byteLength > 32) {
        throw new Error('Only 32 byte integers is supported');
    }

    const uint8Array = new Uint8Array(arrayBuffer);
    let integer = 0;
    let byteCount = 0;

    do {
        /* eslint-disable no-bitwise */
        integer = (integer << 8) + uint8Array[byteCount];
        /* eslint-enable no-bitwise */
        byteCount += 1;
    } while (arrayBuffer.byteLength > byteCount);

    return integer;
};

