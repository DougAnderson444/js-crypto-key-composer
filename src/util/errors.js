class BaseError extends Error {
    constructor(message, name, code, props) {
        super(message);
        Error.captureStackTrace(this, this.constructor);
        this.name = name || 'BaseError';

        if (code) {
            this.code = code;
        }

        Object.assign(this, props);
    }
}

export class UnexpectedTypeError extends BaseError {
    constructor(message, props) {
        super(message, 'UnexpectedTypeError', 'UNEXPECTED_TYPE', props);
    }
}

export class InvalidInputKeyError extends BaseError {
    constructor(message, props) {
        super(message, 'InvalidInputKeyError', 'INVALID_INPUT_KEY', props);
    }
}

export class AggregatedInvalidInputKeyError extends BaseError {
    constructor(errors, props) {
        super('No format was able to recognize the input key', 'AggregatedInvalidInputKeyError', 'AGGREGATED_INVALID_INPUT_KEY', { ...props, errors });
    }
}

export class UnsupportedFormatError extends BaseError {
    constructor(format, props) {
        super(`Unsupported format '${format}'`, 'UnsupportedFormatError', 'UNSUPPORTED_FORMAT', props);
    }
}

export class UnsupportedAlgorithmError extends BaseError {
    constructor(message, props) {
        super(message, 'UnsupportedAlgorithmError', 'UNSUPPORTED_ALGORITHM', props);
    }
}

export class InvalidKeyDataError extends BaseError {
    constructor(message, props) {
        super(message, 'InvalidKeyDataError', 'INVALID_KEY_DATA', props);
    }
}

export class MissingPasswordError extends BaseError {
    constructor(message, props) {
        super(message, 'MissingPasswordError', 'MISSING_PASSWORD', props);
    }
}

export class DecryptionFailedError extends BaseError {
    constructor(message, props) {
        super(message, 'DecryptionFailedError', 'DECRYPTION_FAILED', props);
    }
}

export class DecodeAsn1FailedError extends BaseError {
    constructor(modelName, props) {
        super(`Failed to decode ${modelName}`, 'DecodeAsn1FailedError', 'DECODE_ASN1_FAILED', { ...props, modelName });
    }
}

export class EncodeAsn1FailedError extends BaseError {
    constructor(modelName, props) {
        super(`Failed to encode ${modelName}`, 'EncodeAsn1FailedError', 'ENCODE_ASN1_FAILED', props);
    }
}
