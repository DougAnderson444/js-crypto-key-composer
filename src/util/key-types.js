const KEY_TYPES = {
    // RSA key types
    'rsa-encryption': 'rsa',
    'md2-with-rsa-encryption': 'rsa',
    'md4-with-rsa-encryption': 'rsa',
    'md5-with-rsa-encryption': 'rsa',
    'sha1-with-rsa-encryption': 'rsa',
    'sha224-with-rsa-encryption': 'rsa',
    'sha256-with-rsa-encryption': 'rsa',
    'sha384-with-rsa-encryption': 'rsa',
    'sha512-with-rsa-encryption': 'rsa',
    'sha512-224-with-rsa-encryption': 'rsa',
    'sha512-256-with-rsa-encryption': 'rsa',
    'rsaes-oaep': 'rsa',
    'rsassa-pss': 'rsa',

    // ED25519 key types
    ed25519: 'ed25519',
};

export const KEY_ALIASES = {
    rsa: { id: 'rsa-encryption' },
};

export default KEY_TYPES;
