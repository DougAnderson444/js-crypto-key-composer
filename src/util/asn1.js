import asn1 from '@lordvlad/asn1.js';
import deepForEach from 'deep-for-each';
import cloneDeep from 'clone-deep';
import { nodeBufferToArrayBuffer, arrayBufferToNodeBuffer, bnToArrayBuffer, binaryStringToArrayBuffer } from './binary';

// Ensure that all asn1 objid are returned as strings separated with '.'
// See https://github.com/indutny/asn1.js/blob/b99ce086320e0123331e6272f6de75548c6855fa/lib/asn1/decoders/der.js#L198
// See https://github.com/indutny/asn1.js/blob/b99ce086320e0123331e6272f6de75548c6855fa/lib/asn1/encoders/der.js#L103
const objIdValues = new Proxy({}, {
    get: (obj, key) => {
        if (key === 'hasOwnProperty') {
            return (key) => key.indexOf('.') > 0;
        }

        return key.indexOf('.') > 0 ? key : undefined;
    },
});

export const decodeAsn1 = (encodedEntity, Model) => {
    let decodedEntity;

    try {
        decodedEntity = Model.decode(arrayBufferToNodeBuffer(encodedEntity), 'der');
    } catch (err) {
        throw Object.assign(
            new Error(`Unable to decode ${Model.name}`),
            { code: 'DECODE_ASN1_FAILED', model: Model.name, originalError: err },
        );
    }

    const mapValue = (value) => {
        // Node buffer to array buffer
        if (value && value.buffer) {
            return nodeBufferToArrayBuffer(value);
        }
        // Big number to array buffer
        if (value && value.toArrayLike) {
            return bnToArrayBuffer(value);
        }

        return value;
    };

    // Apply conversion to all properties deep within the entity
    deepForEach(decodedEntity, (value, key, subject) => {
        subject[key] = mapValue(value);
    });

    return mapValue(decodedEntity);
};

export const encodeAsn1 = (decodedEntity, Model) => {
    const mapValue = (value) => {
        // Array buffer to node buffer
        if (value instanceof ArrayBuffer) {
            return arrayBufferToNodeBuffer(value);
        }

        return value;
    };

    // Clone argument because we are going to mutate it
    decodedEntity = cloneDeep(decodedEntity);

    // Apply conversion to all properties deep within the entity
    decodedEntity = mapValue(decodedEntity);
    deepForEach(decodedEntity, (value, key, subject) => {
        subject[key] = mapValue(value);
    });

    let encodedEntity;

    try {
        encodedEntity = Model.encode(decodedEntity, 'der');
    } catch (err) {
        throw Object.assign(
            new Error(`Unable to encode ${Model.name}`),
            { code: 'ENCODE_ASN1_FAILED', model: Model.name, originalError: err },
        );
    }

    return nodeBufferToArrayBuffer(encodedEntity);
};

/* eslint-disable babel/no-invalid-this, newline-per-chained-call */
export const OtherPrimeInfo = asn1.define('OtherPrimeInfo', function () {
    this.seq().obj(
        this.key('prime').int(),
        this.key('exponent').int(),
        this.key('coefficient').int(),
    );
});

export const RSAPrivateKey = asn1.define('RSAPrivateKey', function () {
    this.seq().obj(
        this.key('version').int(),
        this.key('modulus').int(),
        this.key('publicExponent').int(),
        this.key('privateExponent').int(),
        this.key('prime1').int(),
        this.key('prime2').int(),
        this.key('exponent1').int(),
        this.key('exponent2').int(),
        this.key('coefficient').int(),
        this.key('otherPrimeInfos').seqof(OtherPrimeInfo).optional(),
    );
});

export const AlgorithmIdentifier = asn1.define('AlgorithmIdentifier', function () {
    this.seq().obj(
        this.key('id').objid(objIdValues),
        this.key('parameters').optional().any()
    );
});

export const PublicKeyInfo = asn1.define('PublicKeyInfo', function () {
    this.seq().obj(
        this.key('algorithm').use(AlgorithmIdentifier),
        this.key('publicKey').octstr()
    );
});

export const PrivateKeyInfo = asn1.define('PrivateKeyInfo', function () {
    this.seq().obj(
        this.key('version').int(),
        this.key('algorithm').use(AlgorithmIdentifier),
        this.key('privateKey').octstr(),
        this.key('attributes').implicit(0).optional().any(),
        this.key('publicKey').implicit(1).optional().bitstr()
    );
});

export const EncryptedPrivateKeyInfo = asn1.define('EncryptedPrivateKeyInfo', function () {
    this.seq().obj(
        this.key('encryptionAlgorithm').use(AlgorithmIdentifier),
        this.key('encryptedData').octstr(),
    );
});

export const PBES2Algorithms = asn1.define('PBES2Algorithms', function () {
    this.seq().obj(
        this.key('keyDerivationFunc').use(AlgorithmIdentifier),
        this.key('encryptionScheme').use(AlgorithmIdentifier).def({
            id: '1.2.840.113549.2.7', // hmacWithSHA1
            parameters: arrayBufferToNodeBuffer(binaryStringToArrayBuffer('0500')),
        })
    );
});

export const PBES2ESParams = {
    desCB: asn1.define('desCB', function () { this.octstr(); }),
    'des-EDE3-CBC': asn1.define('des-EDE3-CBC', function () { this.octstr(); }),
    'aes128-CBC': asn1.define('aes128-CBC', function () { this.octstr(); }),
    'aes192-CBC': asn1.define('aes192-CBC', function () { this.octstr(); }),
    'aes256-CBC': asn1.define('aes256-CBC', function () { this.octstr(); }),
};

export const PBKDF2params = asn1.define('PBKDF2-params', function () {
    this.seq().obj(
        this.key('salt').octstr(),
        this.key('iterationCount').int(),
        this.key('keyLength').int().optional(),
        this.key('prf').use(AlgorithmIdentifier)
    );
});

export const CurvePrivateKey = asn1.define('CurvePrivateKey', function () { this.octstr(); });
/* eslint-enable babel/no-invalid-this, newline-per-chained-call */
