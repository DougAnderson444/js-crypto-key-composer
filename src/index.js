import { pkcs1Der, pkcs1Pem, pkcs8Der, pkcs8Pem, spkiDer } from './formats';
import { validateInputKey, validateFormat, validateDecomposedKey } from './util/validator';
import { AggregatedInvalidInputKeyError, InvalidInputKeyError } from './util/errors';
import KEY_TYPES from './util/key-types';

const PRIVATE_FORMATS = {
    'pkcs1-pem': pkcs1Pem,
    'pkcs8-pem': pkcs8Pem,
    'pkcs1-der': pkcs1Der,
    'pkcs8-der': pkcs8Der,
};

const PUBLIC_FORMATS = {
    'spki-der': spkiDer,
};

const decomposeKey = (supportedFormats, inputKey, options) => {
    inputKey = validateInputKey(inputKey);
    options = {
        password: null,
        format: Object.keys(supportedFormats),
        ...options,
    };

    if (!Array.isArray(options.format)) {
        const format = validateFormat(options.format, supportedFormats);

        return supportedFormats[format].decomposeKey(inputKey, options);
    }

    const formats = options.format.map((format) => validateFormat(format, supportedFormats));

    // Attempt to decompose the keys, until one succeeds
    // Along the way, we collect the errors for each format
    const errors = {};
    let decomposeKey;

    for (let x = 0; x < formats.length; x += 1) {
        const format = formats[x];

        try {
            decomposeKey = supportedFormats[format].decomposeKey(inputKey, options);
            break;
        } catch (err) {
            // Skip if the error is a InvalidInputKeyError
            if (err instanceof InvalidInputKeyError) {
                errors[format] = err;
                continue;
            }

            err.format = format;
            throw err;
        }
    }

    if (!decomposeKey) {
        throw new AggregatedInvalidInputKeyError(errors);
    }

    return decomposeKey;
};

const composeKey = (supportedFormats, decomposedKey, options) => {
    options = {
        password: null,
        ...options,
    };

    decomposedKey = validateDecomposedKey(decomposedKey, supportedFormats);

    return supportedFormats[decomposedKey.format].composeKey(decomposedKey, options);
};

export const decomposePrivateKey = (inputKey, options) => decomposeKey(PRIVATE_FORMATS, inputKey, options);

export const composePrivateKey = (decomposedKey, options) => composeKey(PRIVATE_FORMATS, decomposedKey, options);

export const decomposePublicKey = (inputKey, options) => decomposeKey(PUBLIC_FORMATS, inputKey, options);

export const composePublicKey = (decomposedKey, options) => composeKey(PUBLIC_FORMATS, decomposedKey, options);

export const getKeyTypeFromAlgorithm = (keyAlgorithm) => {
    const keyAlgorithmId = typeof keyAlgorithm === 'string' ? keyAlgorithm : keyAlgorithm && keyAlgorithm.id;

    return KEY_TYPES[keyAlgorithmId];
};
