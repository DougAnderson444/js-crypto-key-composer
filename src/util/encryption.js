import { createBuffer } from 'node-forge/lib/util';
import sha1 from 'node-forge/lib/sha1';
import sha256 from 'node-forge/lib/sha256';
import sha512 from 'node-forge/lib/sha512';
import aes from 'node-forge/lib/aes';
import des from 'node-forge/lib/des';
import pbkdf2 from 'node-forge/lib/pbkdf2';
import { OIDS } from '../util/oid';
import { arrayBufferToBinaryString, binaryStringToArrayBuffer } from '../util/binary';

export const createPbkdf2 = (params) => {
    const { salt, iterationCount, keyLength, prf } = params;

    const saltStr = arrayBufferToBinaryString(salt);
    const prfName = OIDS[params.prf] || prf;
    let prfMd;

    switch (prfName) {
    case 'hmacWithSHA1':
        prfMd = sha1.create();
        break;
    case 'hmacWithSHA224':
    case 'hmacWithSHA256':
        prfMd = sha256.create();
        break;
    case 'hmacWithSHA384':
    case 'hmacWithSHA512':
        prfMd = sha512.create();
        break;
    default:
        throw Object.assign(
            new Error(`Unsupported prf id '${prf}'`),
            { code: 'UNSUPPORTED_ALGORITHM' }
        );
    }

    return (password) => binaryStringToArrayBuffer(pbkdf2(password, saltStr, iterationCount, keyLength, prfMd));
};

export const createAesDecrypter = (params) => {
    const { iv, mode } = params;

    const ivStr = arrayBufferToBinaryString(iv);

    return (key, encryptedData) => {
        const keyStr = arrayBufferToBinaryString(key);
        const cipher = aes.createDecryptionCipher(keyStr, mode);

        cipher.start(ivStr);
        cipher.update(createBuffer(arrayBufferToBinaryString(encryptedData)));

        if (!cipher.finish()) {
            throw Object.assign(
                new Error('Decryption failed, mostly likely the password is wrong'),
                { code: 'DECRYPTION_FAILED' }
            );
        }

        return binaryStringToArrayBuffer(cipher.output.getBytes());
    };
};

export const createAesEncrypter = (params) => {
    const { iv, mode } = params;

    const ivStr = arrayBufferToBinaryString(iv);

    return (key, data) => {
        const keyStr = arrayBufferToBinaryString(key);
        const cipher = aes.createEncryptionCipher(keyStr, mode);

        cipher.start(ivStr);
        cipher.update(createBuffer(arrayBufferToBinaryString(data)));
        cipher.finish();

        return binaryStringToArrayBuffer(cipher.output.getBytes());
    };
};

export const createDesDecrypter = (params) => {
    const { iv, mode } = params;

    const ivStr = arrayBufferToBinaryString(iv);

    return (key, encryptedData) => {
        const cipher = des.createDecryptionCipher(key, mode);

        cipher.start(ivStr);
        cipher.update(createBuffer(arrayBufferToBinaryString(encryptedData)));

        if (!cipher.finish()) {
            throw Object.assign(
                new Error('Decryption failed, mostly likely the password is wrong'),
                { code: 'DECRYPTION_FAILED' }
            );
        }

        return binaryStringToArrayBuffer(cipher.output.getBytes());
    };
};

export const createDesEncrypter = (params) => {
    const { iv, mode } = params;

    const ivStr = arrayBufferToBinaryString(iv);

    return (key, data) => {
        const keyStr = arrayBufferToBinaryString(key);
        const cipher = des.createEncryptionCipher(keyStr, mode);

        cipher.start(ivStr);
        cipher.update(createBuffer(arrayBufferToBinaryString(data)));
        cipher.finish();

        return binaryStringToArrayBuffer(cipher.output.getBytes());
    };
};
