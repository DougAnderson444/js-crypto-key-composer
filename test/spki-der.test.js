import fs from 'fs';
import { decomposePublicKey, composePublicKey } from '../src';
import { typedArrayToUint8Array } from '../src/util/binary';

const KEYS = {
    'rsa-1': fs.readFileSync('test/fixtures/spki-der/rsa-1'),
    'ed25519-1': fs.readFileSync('test/fixtures/spki-der/ed25519-1'),
    'invalid-1': fs.readFileSync('test/fixtures/spki-der/invalid-1'),
};

describe('decomposePublicKey', () => {
    it('should decompose a standard RSA key', () => {
        expect(decomposePublicKey(KEYS['rsa-1'], { format: 'spki-der' })).toMatchSnapshot();
    });

    it('should decompose a ed25519 key', () => {
        expect(decomposePublicKey(KEYS['ed25519-1'], { format: 'spki-der' })).toMatchSnapshot();
    });

    it('should also support Uint8Array, ArrayBuffer and string besides Node\'s Buffer', () => {
        const nodeBuffer = fs.readFileSync('test/fixtures/spki-der/rsa-1');

        expect(decomposePublicKey(typedArrayToUint8Array(nodeBuffer), { format: 'spki-der' })).toMatchSnapshot();
        expect(decomposePublicKey(typedArrayToUint8Array(nodeBuffer).buffer, { format: 'spki-der' })).toMatchSnapshot();
        expect(decomposePublicKey(nodeBuffer.toString('binary'), { format: 'spki-der' })).toMatchSnapshot();
    });

    it('should fail if the key algorithm is not supported', () => {
        expect.assertions(2);

        try {
            decomposePublicKey(KEYS['invalid-1'], { format: 'spki-der' });
        } catch (err) {
            expect(err.message).toBe('Unsupported key algorithm OID \'0.20.999\'');
            expect(err.code).toBe('UNSUPPORTED_ALGORITHM');
        }
    });

    it('should fail if the input key is invalid', () => {
        expect.assertions(2);

        try {
            decomposePublicKey('', { format: 'spki-der' });
        } catch (err) {
            expect(err.message).toBe('Failed to decode SubjectPublicKeyInfo');
            expect(err.code).toBe('INVALID_INPUT_KEY');
        }
    });
});

describe('composePublicKey', () => {
    it('should compose a standard RSA key (mirroring)', () => {
        const decomposedKey = decomposePublicKey(KEYS['rsa-1'], { format: 'spki-der' });
        const composedKey = composePublicKey(decomposedKey);

        expect(composedKey).toEqual(typedArrayToUint8Array(KEYS['rsa-1']));
    });

    it('should compose a ed25519 key (mirroring)', () => {
        const decomposedKey = decomposePublicKey(KEYS['ed25519-1'], { format: 'spki-der' });
        const composedKey = composePublicKey(decomposedKey);

        expect(composedKey).toEqual(typedArrayToUint8Array(KEYS['ed25519-1']));
    });

    it('should fail if the key algorithm is not supported', () => {
        expect.assertions(2);

        try {
            composePublicKey({
                format: 'spki-der',
                keyAlgorithm: { id: 'foo' },
                keyData: {},
            });
        } catch (err) {
            expect(err.message).toBe('Unsupported key algorithm id \'foo\'');
            expect(err.code).toBe('UNSUPPORTED_ALGORITHM');
        }
    });

    it('should support a string in the key algorithm', () => {
        const decomposedKey = decomposePublicKey(KEYS['rsa-1'], { format: 'spki-der' });
        const composedKey = composePublicKey({ ...decomposedKey, keyAlgorithm: 'rsa-encryption' });

        expect(composedKey).toEqual(typedArrayToUint8Array(KEYS['rsa-1']));
    });

    it('should support the \'rsa\' alias as the key algorithm', () => {
        const decomposedKey = decomposePublicKey(KEYS['rsa-1'], { format: 'spki-der' });
        const composedKey = composePublicKey({ ...decomposedKey, keyAlgorithm: 'rsa' });

        expect(composedKey).toEqual(typedArrayToUint8Array(KEYS['rsa-1']));
    });
});
