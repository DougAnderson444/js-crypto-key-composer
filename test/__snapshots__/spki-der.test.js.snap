// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decomposePublicKey should also support Uint8Array, ArrayBuffer and string besides Node's Buffer 1`] = `
Object {
  "format": "spki-der",
  "keyAlgorithm": Object {
    "id": "rsa-encryption",
  },
  "keyData": Object {
    "modulus": Uint8Array [
      0,
      166,
      53,
      131,
      232,
      151,
      254,
      22,
      219,
      180,
      187,
      248,
      106,
      17,
      27,
      154,
      231,
      54,
      49,
      36,
      22,
      10,
      198,
      20,
      164,
      30,
      244,
      55,
      252,
      5,
      253,
      91,
      131,
      90,
      129,
      232,
      148,
      26,
      193,
      236,
      165,
      71,
      201,
      95,
      139,
      165,
      207,
      106,
      0,
      89,
      83,
      231,
      68,
      86,
      202,
      67,
      242,
      193,
      62,
      0,
      101,
      111,
      4,
      4,
      197,
      136,
      152,
      220,
      168,
      57,
      214,
      225,
      203,
      42,
      253,
      214,
      225,
      142,
      179,
      82,
      91,
      227,
      242,
      13,
      96,
      98,
      232,
      120,
      169,
      74,
      249,
      197,
      192,
      205,
      62,
      17,
      16,
      110,
      162,
      125,
      134,
      42,
      65,
      235,
      199,
      202,
      41,
      193,
      216,
      79,
      100,
      159,
      69,
      241,
      45,
      222,
      131,
      197,
      102,
      222,
      131,
      59,
      53,
      178,
      219,
      77,
      171,
      215,
      148,
      229,
      144,
      181,
      19,
      37,
      50,
      142,
      193,
      34,
      125,
      168,
      105,
      124,
      205,
      161,
      248,
      237,
      32,
      202,
      94,
      83,
      170,
      95,
      67,
      147,
      177,
      80,
      30,
      57,
      208,
      36,
      47,
      199,
      13,
      173,
      241,
      116,
      71,
      78,
      196,
      191,
      155,
      252,
      169,
      45,
      103,
      252,
      95,
      115,
      168,
      150,
      162,
      222,
      90,
      59,
      204,
      249,
      122,
      79,
      65,
      147,
      255,
      57,
      159,
      18,
      15,
      215,
      26,
      167,
      253,
      146,
      210,
      25,
      66,
      88,
      127,
      80,
      8,
      218,
      207,
      54,
      29,
      81,
      62,
      87,
      148,
      139,
      188,
      114,
      54,
      229,
      199,
      62,
      227,
      63,
      169,
      42,
      0,
      93,
      168,
      174,
      255,
      38,
      147,
      1,
      41,
      44,
      83,
      215,
      89,
      81,
      49,
      60,
      228,
      54,
      249,
      107,
      156,
      177,
      168,
      50,
      231,
      81,
      155,
      171,
      209,
      190,
      213,
    ],
    "publicExponent": 65537,
  },
}
`;

exports[`decomposePublicKey should also support Uint8Array, ArrayBuffer and string besides Node's Buffer 2`] = `
Object {
  "format": "spki-der",
  "keyAlgorithm": Object {
    "id": "rsa-encryption",
  },
  "keyData": Object {
    "modulus": Uint8Array [
      0,
      166,
      53,
      131,
      232,
      151,
      254,
      22,
      219,
      180,
      187,
      248,
      106,
      17,
      27,
      154,
      231,
      54,
      49,
      36,
      22,
      10,
      198,
      20,
      164,
      30,
      244,
      55,
      252,
      5,
      253,
      91,
      131,
      90,
      129,
      232,
      148,
      26,
      193,
      236,
      165,
      71,
      201,
      95,
      139,
      165,
      207,
      106,
      0,
      89,
      83,
      231,
      68,
      86,
      202,
      67,
      242,
      193,
      62,
      0,
      101,
      111,
      4,
      4,
      197,
      136,
      152,
      220,
      168,
      57,
      214,
      225,
      203,
      42,
      253,
      214,
      225,
      142,
      179,
      82,
      91,
      227,
      242,
      13,
      96,
      98,
      232,
      120,
      169,
      74,
      249,
      197,
      192,
      205,
      62,
      17,
      16,
      110,
      162,
      125,
      134,
      42,
      65,
      235,
      199,
      202,
      41,
      193,
      216,
      79,
      100,
      159,
      69,
      241,
      45,
      222,
      131,
      197,
      102,
      222,
      131,
      59,
      53,
      178,
      219,
      77,
      171,
      215,
      148,
      229,
      144,
      181,
      19,
      37,
      50,
      142,
      193,
      34,
      125,
      168,
      105,
      124,
      205,
      161,
      248,
      237,
      32,
      202,
      94,
      83,
      170,
      95,
      67,
      147,
      177,
      80,
      30,
      57,
      208,
      36,
      47,
      199,
      13,
      173,
      241,
      116,
      71,
      78,
      196,
      191,
      155,
      252,
      169,
      45,
      103,
      252,
      95,
      115,
      168,
      150,
      162,
      222,
      90,
      59,
      204,
      249,
      122,
      79,
      65,
      147,
      255,
      57,
      159,
      18,
      15,
      215,
      26,
      167,
      253,
      146,
      210,
      25,
      66,
      88,
      127,
      80,
      8,
      218,
      207,
      54,
      29,
      81,
      62,
      87,
      148,
      139,
      188,
      114,
      54,
      229,
      199,
      62,
      227,
      63,
      169,
      42,
      0,
      93,
      168,
      174,
      255,
      38,
      147,
      1,
      41,
      44,
      83,
      215,
      89,
      81,
      49,
      60,
      228,
      54,
      249,
      107,
      156,
      177,
      168,
      50,
      231,
      81,
      155,
      171,
      209,
      190,
      213,
    ],
    "publicExponent": 65537,
  },
}
`;

exports[`decomposePublicKey should also support Uint8Array, ArrayBuffer and string besides Node's Buffer 3`] = `
Object {
  "format": "spki-der",
  "keyAlgorithm": Object {
    "id": "rsa-encryption",
  },
  "keyData": Object {
    "modulus": Uint8Array [
      0,
      166,
      53,
      131,
      232,
      151,
      254,
      22,
      219,
      180,
      187,
      248,
      106,
      17,
      27,
      154,
      231,
      54,
      49,
      36,
      22,
      10,
      198,
      20,
      164,
      30,
      244,
      55,
      252,
      5,
      253,
      91,
      131,
      90,
      129,
      232,
      148,
      26,
      193,
      236,
      165,
      71,
      201,
      95,
      139,
      165,
      207,
      106,
      0,
      89,
      83,
      231,
      68,
      86,
      202,
      67,
      242,
      193,
      62,
      0,
      101,
      111,
      4,
      4,
      197,
      136,
      152,
      220,
      168,
      57,
      214,
      225,
      203,
      42,
      253,
      214,
      225,
      142,
      179,
      82,
      91,
      227,
      242,
      13,
      96,
      98,
      232,
      120,
      169,
      74,
      249,
      197,
      192,
      205,
      62,
      17,
      16,
      110,
      162,
      125,
      134,
      42,
      65,
      235,
      199,
      202,
      41,
      193,
      216,
      79,
      100,
      159,
      69,
      241,
      45,
      222,
      131,
      197,
      102,
      222,
      131,
      59,
      53,
      178,
      219,
      77,
      171,
      215,
      148,
      229,
      144,
      181,
      19,
      37,
      50,
      142,
      193,
      34,
      125,
      168,
      105,
      124,
      205,
      161,
      248,
      237,
      32,
      202,
      94,
      83,
      170,
      95,
      67,
      147,
      177,
      80,
      30,
      57,
      208,
      36,
      47,
      199,
      13,
      173,
      241,
      116,
      71,
      78,
      196,
      191,
      155,
      252,
      169,
      45,
      103,
      252,
      95,
      115,
      168,
      150,
      162,
      222,
      90,
      59,
      204,
      249,
      122,
      79,
      65,
      147,
      255,
      57,
      159,
      18,
      15,
      215,
      26,
      167,
      253,
      146,
      210,
      25,
      66,
      88,
      127,
      80,
      8,
      218,
      207,
      54,
      29,
      81,
      62,
      87,
      148,
      139,
      188,
      114,
      54,
      229,
      199,
      62,
      227,
      63,
      169,
      42,
      0,
      93,
      168,
      174,
      255,
      38,
      147,
      1,
      41,
      44,
      83,
      215,
      89,
      81,
      49,
      60,
      228,
      54,
      249,
      107,
      156,
      177,
      168,
      50,
      231,
      81,
      155,
      171,
      209,
      190,
      213,
    ],
    "publicExponent": 65537,
  },
}
`;

exports[`decomposePublicKey should decompose a ed25519 key 1`] = `
Object {
  "format": "spki-der",
  "keyAlgorithm": Object {
    "id": "ed25519",
  },
  "keyData": Object {
    "bytes": Uint8Array [
      44,
      164,
      164,
      151,
      38,
      69,
      162,
      156,
      247,
      237,
      98,
      184,
      38,
      41,
      134,
      207,
      193,
      28,
      252,
      42,
      53,
      179,
      38,
      69,
      172,
      88,
      192,
      91,
      227,
      191,
      189,
      60,
    ],
  },
}
`;

exports[`decomposePublicKey should decompose a standard RSA key 1`] = `
Object {
  "format": "spki-der",
  "keyAlgorithm": Object {
    "id": "rsa-encryption",
  },
  "keyData": Object {
    "modulus": Uint8Array [
      0,
      166,
      53,
      131,
      232,
      151,
      254,
      22,
      219,
      180,
      187,
      248,
      106,
      17,
      27,
      154,
      231,
      54,
      49,
      36,
      22,
      10,
      198,
      20,
      164,
      30,
      244,
      55,
      252,
      5,
      253,
      91,
      131,
      90,
      129,
      232,
      148,
      26,
      193,
      236,
      165,
      71,
      201,
      95,
      139,
      165,
      207,
      106,
      0,
      89,
      83,
      231,
      68,
      86,
      202,
      67,
      242,
      193,
      62,
      0,
      101,
      111,
      4,
      4,
      197,
      136,
      152,
      220,
      168,
      57,
      214,
      225,
      203,
      42,
      253,
      214,
      225,
      142,
      179,
      82,
      91,
      227,
      242,
      13,
      96,
      98,
      232,
      120,
      169,
      74,
      249,
      197,
      192,
      205,
      62,
      17,
      16,
      110,
      162,
      125,
      134,
      42,
      65,
      235,
      199,
      202,
      41,
      193,
      216,
      79,
      100,
      159,
      69,
      241,
      45,
      222,
      131,
      197,
      102,
      222,
      131,
      59,
      53,
      178,
      219,
      77,
      171,
      215,
      148,
      229,
      144,
      181,
      19,
      37,
      50,
      142,
      193,
      34,
      125,
      168,
      105,
      124,
      205,
      161,
      248,
      237,
      32,
      202,
      94,
      83,
      170,
      95,
      67,
      147,
      177,
      80,
      30,
      57,
      208,
      36,
      47,
      199,
      13,
      173,
      241,
      116,
      71,
      78,
      196,
      191,
      155,
      252,
      169,
      45,
      103,
      252,
      95,
      115,
      168,
      150,
      162,
      222,
      90,
      59,
      204,
      249,
      122,
      79,
      65,
      147,
      255,
      57,
      159,
      18,
      15,
      215,
      26,
      167,
      253,
      146,
      210,
      25,
      66,
      88,
      127,
      80,
      8,
      218,
      207,
      54,
      29,
      81,
      62,
      87,
      148,
      139,
      188,
      114,
      54,
      229,
      199,
      62,
      227,
      63,
      169,
      42,
      0,
      93,
      168,
      174,
      255,
      38,
      147,
      1,
      41,
      44,
      83,
      215,
      89,
      81,
      49,
      60,
      228,
      54,
      249,
      107,
      156,
      177,
      168,
      50,
      231,
      81,
      155,
      171,
      209,
      190,
      213,
    ],
    "publicExponent": 65537,
  },
}
`;
